torchbiggraph/examples/e2e/data_prep.py

```diff
@@ -0,0 +1,325 @@
+    rels = pd.read_sql_query(query, conn)
+    end = time.time()
+    logging.info(f"Remapped relationships in {end - start}s")
+    return rels
+
+
+def remap_entities(conn, entity2partitions):
+    """
+    A function to remap entities with partitioning using SQL queries.
+
+    This function is complicated because the partitions have to be
+    constructed first, and then we have to generate ordinal mappings of
+    entity ids. These mappings will be used to generate buckets of edges
+    for training and then for mapping our edges back to their original
+    ids for use in downstream tasks.
+    """
+    logging.info("Remapping entities")
+    start=time.time()
+    query = ""
+    for entity, npartitions in entity2partitions.items():
+        query = type_tmp_table.format(type=entity, nparts=npartitions)
+
+        for i in range(npartitions):
+            query += partitioned_mapped_entities.format(type=entity, n=i)
+        logging.debug(f"Running query: {query}")
+        conn.executescript(query)
+    end = time.time()
+    logging.info(f"Remapped entities in {end - start}s")
+
+
+def generate_ctes(lhs_part, rhs_part, rels, entity2partitions):
+    """
+    This function generates the sub-table CTEs that help us generate
+    the completed edgelist.
+    """
+    nctes = 0
+    ctes = """
+    with cte_0 as (
+    """
+    first = True
+    for _ , r in rels.iterrows():
+        if lhs_part >= entity2partitions[r['source_type']]:
+            continue
+        if rhs_part >= entity2partitions[r['destination_type']]:
+            continue
+        if not first:
+            ctes += f", cte_{nctes} as ("
+        ctes += edgelist_cte_mapper.format(
+            rel_name=r['id'],
+            lhs_type=r['source_type'],
+            rhs_type=r['destination_type'],
+            i = lhs_part,
+            j = rhs_part,
+        )
+        ctes += ")"
+
+        nctes += 1
+        first = False
+    return nctes, ctes
+
+
+def generate_unions(nctes):
+    """
+    This function is just a helper function for
+    generating the final edge list tables.
+    """
+    subquery = ""
+    first = True
+    for i in range(nctes):
+        if not first:
+            subquery += "\tunion\n"
+        subquery += f"\tselect * from cte_{i}\n"
+        first = False
+    return subquery
+
+
+def remap_edges(conn, rels, entity2partitions):
+    """
+    A function to remap all edges to ordinal IDs
+    according to their type.
+    """
+    logging.info("Remapping edges")
+    start = time.time()
+
+    nentities_premap = conn.execute("""
+    select count(*) from edges
+    ;
+    """).fetchall()[0][0]
+
+    query = ""
+    NPARTS = max(entity2partitions.values())
+    for lhs_part in range(NPARTS):
+        for rhs_part in range(NPARTS):
+            nctes, ctes = generate_ctes(lhs_part, rhs_part, rels, entity2partitions)
+            subquery = generate_unions(nctes)
+            query += edges_partitioned.format(
```

Comment:
So you create a separate table for each edge type? 

It seems a bit dangerous that you're dropping generically named tables from the db.
1. At least prefix with `tmp` so we don't delete anything important
2. Gotta be careful if there are multiple people running the script at a tie on the same db.

Reply:
Oh, and I create a different CTE for each edge-type. Because I'm only calling the CTE once, it's largely just a way to visually group the work into finite, easy to understand, work packages for debugging